field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -0,0 +1,382 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-| TypeChecker
+
+We provide a simple bidirectional typechecker
+
+Potential resources:
+* https://www.cs.cmu.edu/~fp/courses/15312-f04/handouts/15-bidirectional.pdf
+* http://www.cse.chalmers.se/~ulfn/papers/afp08/tutorial.pdf
+* http://www.davidchristiansen.dk/tutorials/bidirectional.pdf
+-}
+
+module TinyLang.Field.Typed.TypeChecker
+    ( TypeCheckError
+    , MonadTypeError
+    , MonadTypeChecker
+    , TypeCheckerT(..)
+    , TypeChecker
+    , runTypeChecker
+    , typeCheck
+    , checkType
+    , inferExpr
+    , checkExpr
+    , inferUniVar
+    , checkUniVar
+    )
+    where
+
+import           TinyLang.Prelude hiding (TypeError)
+
+import           Data.Field
+import           TinyLang.Environment
+import           TinyLang.Var
+import           TinyLang.ParseUtils
+import           TinyLang.Field.Evaluator
+import           TinyLang.Field.UniConst
+import           TinyLang.Field.Existential
+import qualified TinyLang.Field.Typed.Core  as T
+import qualified TinyLang.Field.Raw.Core    as R
+
+import           Data.Kind
+import qualified Data.String.Interpolate    as QQ
+
+{-| Utility Type Aliases or Constraints
+-}
+type TypeCheckError     = String
+type MonadTypeError   m = MonadError TypeCheckError m
+type MonadTypeChecker m = ( MonadSupply m
+                          , MonadScope m
+                          , MonadTypeError m
+                          )
+
+{-| TypeCheckerT Transformer
+-}
+newtype TypeCheckerT e (m :: Type -> Type) a =
+    TypeChecker { runTypeCheckerT :: (ExceptT e (StateT Scope (SupplyT m))) a }
+    deriving newtype ( Monad
+                     , Functor
+                     , Applicative
+                     , MonadError e
+                     , MonadSupply
+                     , MonadScope
+                     )
+
+{-| A simple type checker
+-}
+type TypeChecker = TypeCheckerT TypeCheckError Identity
+
+{-|
+-}
+runTypeChecker :: TypeChecker a -> Either TypeCheckError a
+runTypeChecker typeChecker =
+    runIdentity $ runSupplyT
+                $ evalStateT (runExceptT (runTypeCheckerT typeChecker))
+                             mempty
+
+{-|
+-}
+typeCheck
+    :: forall f. (TextField f)
+    => R.Expr R.Var f -> Either TypeCheckError (T.SomeUniExpr f)
+typeCheck = runTypeChecker . inferExpr
+
+{-|
+-}
+checkType
+    :: forall f a. (TextField f, KnownUni f a)
+    => R.Expr R.Var f -> Either TypeCheckError (T.Expr f a)
+checkType = runTypeChecker . checkExpr
+
+
+{-| = Bidirectional Typing Rules
+
+Here is a quick recap of our current syntax of the language
```

Comment:
We'll be changing syntax and keeping the grammar rules up-to-date is going to be annoying. I think the notation is readable without them being inlined here, we can refer the reader to the module where those rules are introduced.

I'm not against having the syntax described here, just worried it might be expensive maintenance-wise to keep those docs up-to-date.

Reply:
Ok, I will drop most of the comments then.